File: packages/core-sdk/src/modules/memory.ts
import type { HttpClient } from "../client/http-client.js";

export interface MemoryTargetScope {
  teamId?: number;
  runtimeType?: string | null;
  deviceId?: string | number | null;
  agentId?: string | null;
}

export interface MemoryStoreRequest extends MemoryTargetScope {
  content: string;
  category?: "preference" | "fact" | "decision" | "entity" | "other";
  importance?: number;
  sourceSession?: string | null;
  skipDuplicateCheck?: boolean;
}

export interface MemorySearchRequest extends MemoryTargetScope {
  query: string;
  limit?: number;
  threshold?: number;
}

export interface MemoryDeleteResult {
  deleted: boolean;
}

export interface MemoryClearResult {
  clearedCount: number;
}

export interface MemoryStats {
  [key: string]: unknown;
}

export interface MemoryStatsRequest extends MemoryTargetScope {}

export class MemoryModule {
  constructor(private readonly http: HttpClient) {}

  async store(payload: MemoryStoreRequest): Promise<Record<string, unknown>> {
    const result = await this.http.request<Record<string, unknown>>({
      method: "POST",
      path: "/memory/store",
      body: {
        content: payload.content,
        category: payload.category,
        importance: payload.importance,
        team_id: payload.teamId,
        runtime_type: payload.runtimeType,
        device_id: payload.deviceId,
        agent_id: payload.agentId,
        source_session: payload.sourceSession,
        skip_duplicate_check: payload.skipDuplicateCheck,
      },
    });

    return result.entry && typeof result.entry === "object"
      ? result.entry as Record<string, unknown>
      : result;
  }

  async search(payload: MemorySearchRequest): Promise<Record<string, unknown>[]> {
    const result = await this.http.request<Record<string, unknown>[] | { results?: unknown }>({
      method: "POST",
      path: "/memory/search",
      body: {
        query: payload.query,
        limit: payload.limit,
        threshold: payload.threshold,
        team_id: payload.teamId,
        runtime_type: payload.runtimeType,
        device_id: payload.deviceId,
        agent_id: payload.agentId,
      },
    });

    if (Array.isArray(result)) {
      return result;
    }

    return Array.isArray(result.results)
      ? result.results.filter((item): item is Record<string, unknown> => Boolean(item && typeof item === "object"))
      : [];
  }

  async delete(
    entryId: string,
    scope?: Pick<MemoryTargetScope, "teamId" | "agentId" | "runtimeType">,
  ): Promise<MemoryDeleteResult> {
    const result = await this.http.request<{ deleted?: boolean }>({
      method: "DELETE",
      path: `/memory/${encodeURIComponent(entryId)}`,
      query: {
        team_id: scope?.teamId,
        agent_id: scope?.agentId,
        runtime_type: scope?.runtimeType,
      },
    });

    return { deleted: Boolean(result.deleted) };
  }

  async clear(
    agentId: string,
    scope?: Pick<MemoryTargetScope, "teamId" | "runtimeType">,
  ): Promise<MemoryClearResult> {
    const result = await this.http.request<{ cleared_count?: number }>({
      method: "POST",
      path: "/memory/clear",
      body: {
        agent_id: agentId,
        team_id: scope?.teamId,
        runtime_type: scope?.runtimeType,
      },
    });

    return { clearedCount: result.cleared_count ?? 0 };
  }

  async stats(payload: MemoryStatsRequest = {}): Promise<MemoryStats> {
    return this.http.request<MemoryStats>({
      method: "GET",
      path: "/memory/stats",
      query: {
        team_id: payload.teamId,
        agent_id: payload.agentId,
        runtime_type: payload.runtimeType,
      },
    });
  }
}

File: packages/hermes-bridge/bridge/api/memory.py
"""记忆端点：stats / search / store / delete"""

from __future__ import annotations

import logging

from fastapi import APIRouter, Query
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)
router = APIRouter()


class MemorySearchRequest(BaseModel):
    query: str = Field(default="", description="搜索查询文本")
    limit: int = Field(default=5, ge=1, description="最大返回数")
    threshold: float = Field(default=0.0, ge=0.0, description="最低匹配阈值")
    agent_id: str | None = None
    agent_profile: str | None = None
    team_id: str | int | None = None
    runtime_type: str | None = None
    device_id: str | None = None
    teamId: str | int | None = None
    runtimeType: str | None = None
    agentId: str | None = None


class MemoryStoreRequest(BaseModel):
    content: str = Field(default="", description="记忆内容")
    category: str = Field(default="other", description="记忆分类")
    importance: float = Field(default=0.5, ge=0.0, le=1.0, description="重要程度")
    agent_id: str | None = None
    agent_profile: str | None = None
    team_id: str | int | None = None
    runtime_type: str = Field(default="openclaw", description="运行时类型")
    device_id: str | None = None
    source_session: str | None = None
    skip_duplicate_check: bool = False
    teamId: str | int | None = None
    runtimeType: str | None = None
    agentId: str | None = None


def _build_scope_from_query(
    agent_id: str | None = None,
    agent_profile: str | None = None,
    team_id: str | None = None,
    runtime_type: str | None = None,
    device_id: str | None = None,
) -> dict:
    scope: dict = {}
    effective_agent = agent_id or agent_profile
    if effective_agent:
        scope["agent_id"] = effective_agent
    for key, val in [("team_id", team_id), ("runtime_type", runtime_type), ("device_id", device_id)]:
        if val is not None:
            scope[key] = val
    return scope


@router.get("/memory/stats")
async def memory_stats(
    agent_id: str | None = Query(default=None),
    agent_profile: str | None = Query(default=None),
    team_id: str | None = Query(default=None),
    runtime_type: str | None = Query(default=None),
    device_id: str | None = Query(default=None),
):
    try:
        from memory_store import get_memory_stats

        scope = _build_scope_from_query(
            agent_id=agent_id,
            agent_profile=agent_profile,
            team_id=team_id,
            runtime_type=runtime_type,
            device_id=device_id,
        )
        stats = get_memory_stats(scope=scope)
        return JSONResponse({"success": True, **stats})
    except Exception as exc:
        logger.exception("memory stats 异常")
        return JSONResponse({"error": str(exc)}, status_code=500)


@router.post("/memory/search")
async def memory_search(req: MemorySearchRequest):
    try:
        from memory_store import search_memory

        scope = _build_scope_from_query(
            agent_id=req.agent_id,
            agent_profile=req.agent_profile,
            team_id=req.team_id,
            runtime_type=req.runtime_type,
            device_id=req.device_id,
        )
        results = search_memory(
            query=req.query,
            limit=req.limit,
            threshold=req.threshold,
            scope=scope,
        )
        return JSONResponse({"success": True, "results": results})
    except Exception as exc:
        logger.exception("memory search 异常")
        return JSONResponse({"error": str(exc)}, status_code=500)


@router.post("/memory/store")
async def memory_store(req: MemoryStoreRequest):
    try:
        from memory_store import store_memory

        agent_id = req.agent_profile or req.agent_id
        entry = store_memory(
            content=req.content,
            category=req.category,
            importance=req.importance,
            team_id=req.team_id,
            runtime_type=req.runtime_type,
            device_id=req.device_id,
            agent_id=agent_id,
            source_session=req.source_session,
            skip_duplicate_check=req.skip_duplicate_check,
        )
        return JSONResponse({"success": True, "entry": entry})
    except Exception as exc:
        logger.exception("memory store 异常")
        return JSONResponse({"error": str(exc)}, status_code=500)


@router.delete("/memory/{entry_id}")
async def memory_delete(
    entry_id: str,
    team_id: str | None = Query(default=None),
    runtime_type: str | None = Query(default=None),
    device_id: str | None = Query(default=None),
    agent_id: str | None = Query(default=None),
):
    try:
        from memory_store import delete_memory

        scope: dict = {}
        for key, val in [
            ("team_id", team_id),
            ("runtime_type", runtime_type),
            ("device_id", device_id),
            ("agent_id", agent_id),
        ]:
            if val is not None:
                scope[key] = val

        deleted = delete_memory(entry_id, scope=scope)
        return JSONResponse({"success": deleted, "deleted": deleted})
    except Exception as exc:
        logger.exception("memory delete 异常")
        return JSONResponse({"error": str(exc)}, status_code=500)
